ruby/ql/lib/codeql/ruby/frameworks/Array.qll

```diff
@@ -0,0 +1,2407 @@
+    PartitionSummary() { this = "partition" }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      input = "ArrayElement of Receiver" and
+      output =
+        ["Parameter[0] of BlockArgument", "ArrayElement[?] of ArrayElement[?] of ReturnValue"] and
+      preservesValue = true
+    }
+  }
+
+  private class QuerySummary extends SimpleSummarizedCallable {
+    QuerySummary() { this = ["all?", "any?", "none?", "one?"] }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      input = "ArrayElement of Receiver" and
+      output = "Parameter[0] of BlockArgument" and
+      preservesValue = true
+    }
+  }
+
+  private class RejectSummary extends SimpleSummarizedCallable {
+    RejectSummary() { this = "reject" }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      input = "ArrayElement of Receiver" and
+      output = ["Parameter[0] of BlockArgument", "ArrayElement[?] of ReturnValue"] and
+      preservesValue = true
+    }
+  }
+
+  private class RejectBangSummary extends SimpleSummarizedCallable {
+    RejectBangSummary() { this = "reject!" }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      input = "ArrayElement of Receiver" and
+      output =
+        [
+          "ArrayElement[?] of ReturnValue", "ArrayElement[?] of Receiver",
+          "Parameter[0] of BlockArgument"
+        ] and
+      preservesValue = true
+    }
+
+    override predicate clearsContent(ParameterPosition pos, DataFlow::Content content) {
+      pos.isSelf() and
+      content instanceof DataFlow::Content::ArrayElementContent
+    }
+  }
+
+  private class SelectSummary extends SimpleSummarizedCallable {
+    // `find_all` and `filter` are aliases of `select`.
+    SelectSummary() { this = ["select", "find_all", "filter"] }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      input = "ArrayElement of Receiver" and
+      output = ["Parameter[0] of BlockArgument", "ArrayElement[?] of ReturnValue"] and
+      preservesValue = true
+    }
+  }
+
+  private class SliceBeforeAfterSummary extends SimpleSummarizedCallable {
+    SliceBeforeAfterSummary() { this = ["slice_before", "slice_after"] }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      input = "ArrayElement of Receiver" and
+      output = "Parameter[0] of BlockArgument" and
+      preservesValue = true
+    }
+  }
+
+  private class SliceWhenSummary extends SimpleSummarizedCallable {
+    SliceWhenSummary() { this = "slice_when" }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      input = "ArrayElement of Receiver" and
+      output = ["Parameter[0] of BlockArgument", "Parameter[1] of BlockArgument"] and
+      preservesValue = true
```

Comment:
Same here:
`ArrayElement of Receiver` -> `ArrayElement[?] of ArrayElement[?] of ReturnValue`